File: backend/src/main/java/com/voidGustavoNunes/Accenture/model/Fornecedor.java
package com.voidGustavoNunes.Accenture.model;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;
import com.voidGustavoNunes.Accenture.model.enums.TipoPessoa;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToMany;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import lombok.Data;
import lombok.EqualsAndHashCode;

@Entity
@Data
@EqualsAndHashCode(callSuper = true)
public class Fornecedor extends GenericModel{

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @NotBlank(message = "Tipo é obrigatório")
    @Enumerated(EnumType.STRING)
    private TipoPessoa tipo;
    
    //cpf ou cnpj
    @NotBlank(message = "Identificador é obrigatório")
    @Column(unique = true)
    private String identificador;

    @NotBlank(message = "Nome é obrigatório")
    private String nome;
    
    @NotBlank(message = "Email é obrigatório")
    @Email(message = "Email deve ser válido")
    private String email;
    
    @NotBlank(message = "CEP é obrigatório")
    @Pattern(regexp = "\\d{8}", message = "O CEP deve conter 8 dígitos numéricos")
    private String cep;
    
    @ManyToMany(mappedBy = "fornecedores")
    private Set<Empresa> empresas = new HashSet<>();

    private String rg;
    
    private LocalDate dataNascimento;
    
    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Fornecedor that = (Fornecedor) o;
        return Objects.equals(id, that.id);
    }
}

File: frontend/projeto-accenture/src/app/modules/Fornecedor.ts
import { ID } from "@datorama/akita";
import { Empresa } from "./Empresa";
import { TipoPessoa } from "./enums/TipoPessoa";

export class Fornecedor{
  id?: ID;
  tipo: TipoPessoa;
  identificador: string;
  nome: string;
  email: string;
  cep: string;
  rg?: string;
  dataNascimento?: Date;
  empresas?: Empresa[];

  constructor(
    id: | ID,
    tipo: TipoPessoa,
    identificador: string,
    nome: string,
    email: string,
    cep: string,
    rg?: string,
    dataNascimento?: Date,
    empresas?: Empresa[]
  ) {
    this.id = id;
    this.tipo = tipo;
    this.identificador = identificador;
    this.nome = nome;
    this.email = email;
    this.cep = cep;
    this.rg = rg;
    this.dataNascimento = dataNascimento;
    this.empresas = empresas;
  }
}
